src/geocat/viz/contourf/contourf.py

```diff
@@ -1,81 +1,435 @@
         NCL_Plot.__init__(self, *args, **kwargs)
 
-        # Create plot
-        if kwargs.get('contour_fill') is not False:
-            self.cf = self.ax.contourf(
-                self.data,
-                levels=self.levels,
-                cmap=self.cmap,
-                transform=self.projection,
-                extent=[self.xlim[0], self.xlim[1], self.ylim[0], self.ylim[1]])
+        # Add filled contours and/or contour lines to figure, as specified
+        self._generate_contours()
 
-        if kwargs.get('contour_lines') is not False:
-            self.cl = self.ax.contour(
-                self.data,
-                levels=self.levels,
-                colors='black',
-                alpha=0.8,
-                linewidths=0.4,
-                linestyles='solid',
-                transform=self.projection,
-                extent=[self.xlim[0], self.xlim[1], self.ylim[0], self.ylim[1]])
-
-        self._set_NCL_style(self.ax)
-
-        # call colorbar creation from parent class
-        # set colorbar if specified
-        if self.colorbar is not False and self.colorbar is not 'off':
-            self._add_colorbar(self.cf)
+        # Set figure in NCL style
+        _fig_ax._set_NCL_style(self, self.ax)
+
+        # If contour labels are requested, try to set them on contour lines. If failed, use filled contours
+        if self.draw_contour_labels is True:
+            try:
+                self._add_contour_labels(self.ax,
+                                         self.cl,
+                                         contourlabels=self.contourlabels,
+                                         fontsize=self.contourfontsize,
+                                         background=self.contourbackground)
+            except:
+                if self.contour_fill is not False:
+                    self._add_contour_labels(self.ax,
+                                             self.cf,
+                                             contourlabels=self.contourlabels,
+                                             fontsize=self.contourfontsize,
+                                             background=self.contourbackground)
+
+        # Call colorbar creation from parent class
+        # Set colorbar if specified
+
+        # If not a subplot and add_colorbar and contour_fill is not false, add colorbar
+        if (((self.add_colorbar is not False) and
+             (self.add_colorbar != 'off') and
+             (kwargs.get('contour_fill') is not False) and
+             (self.subplot is None)) or
+                # If subplot, check if in last position in subplot and that add_colorbar is not False and plot
+            ((self.add_colorbar is not False) and
+             (self.subplot[2] == self.subplot[0] * self.subplot[1]) and
+             (self.add_colorbar != "off") and
+             ((self.individual_cb is not True) or
+              (self.ref_fig.individual_cb is not True)) or
+             (self.individual_cb is True))):
```

Comment:
`self.add_colorbar`, `self.subplot`, `self.individual_cb` , and `self.ref_fig` need to be created before this logic block. I know they're meant to be specified as kwargs, but you're never actually pulling them out of the kwargs argument.

Reply:
I believe I pull those out in either _plot_util or _set_up_figure - is that okay or should I have them pulled out also in contour?

Reply:
You know, I'm not sure. I'll look into it.